services/telemetry/base.go

```diff
@@ -0,0 +1,154 @@
+package telemetry
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+// Telemetry exported and unexported fields
+type Service struct {
+	Config
+	ticker     *time.Ticker
+	lastStatus string
+	stopChan   chan bool
+	wg         sync.WaitGroup
+	isRunning  bool
+	lock       sync.Mutex
+}
+
+// Telemetry config
+type Config struct {
+	URL      string `yaml:"url'`
+	UUID     string `yaml:"uuid"`
+	Interval int    `yaml:"interval"`
+}
+
+const (
+	defaultURL      = "https://v.percona.com"
+	defaultUUID     = ""
+	defaultinterval = 24 * 60 * 60
+)
+
+var (
+	stat     = os.Stat
+	readFile = ioutil.ReadFile
+	output   = commandOutput
+	Version  = "1.3.0"
+)
+
+// NewService creates a new telemetry service given a configuration
+func NewService(config *Config) (*Service, error) {
+	service := &Service{
+		Config: *config,
+	}
+	return service, nil
+}
+
+// Start sets a new ticker to collect data every Config.Interval seconds.
+func (s *Service) Start() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.isRunning {
+		return
+	}
+	s.stopChan = make(chan bool)
+	s.isRunning = true
+	s.ticker = time.NewTicker(time.Second * time.Duration(s.Interval))
+	s.wg.Add(1)
+	go func() {
+	LOOP:
+		for {
+			select {
+			case <-s.ticker.C:
+				data, err := s.collectData()
+				payload, err := s.makePayload(data)
+				if err != nil {
+					s.lastStatus = fmt.Sprintf("%v cannot build payload for telemetry info: %s", time.Now(), err.Error())
+					continue
+				}
+				err = s.sendRequest(payload)
+				if err != nil {
+					s.lastStatus = fmt.Sprintf("%v error sending telemetry info: %s", time.Now(), err.Error())
+					continue
+				}
+				s.lastStatus = fmt.Sprintf("%v telemetry data sent.", time.Now())
+			case <-s.stopChan:
+				break LOOP
```

Comment:
Once `s.wg.Done()` is moved to `defer`, we can just `return` there, and drop label.